nixos.xfce: use 'XFCE' as the session name

romildo opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

The desktop session name is used for the `DesktopNames` field in the generated session file (`xfce.desktop`). The standard desktop name for Xfce is `XFCE` (all in caps). The value for the `XDG_CURRENT_DESKTOP` environment variable is obtained from this field. When this variable does not have the right value, some features do not work correctly. For instance, menu items set to show only on Xfce are not shown.

Fixes https://github.com/NixOS/nixpkgs/issues/82571

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

romildo committed:
nixos.xfce: use 'XFCE' as the session name

The desktop session name is used for the 'DesktopNames' field in the
generated session file (xfce.desktop). The standard desktop name for
Xfce is 'XFCE' (all in caps). The value for the 'XDG_CURRENT_DESKTOP'
environment variable is obtained from this field. When this variable
does not have the right value, some features do not work
correctly. For instance, menu items set to show only on Xfce are not
shown.

symphorien commented:
Does it have an impact on the correct values of `services.xserver.displayManager.defaultSession` ?
If so relase notes should be updated to be upper case.

romildo commented:
> Does it have an impact on the correct values of `services.xserver.displayManager.defaultSession` ?

Yes, it does. I have tested it.

The original issue could be fixed in three ways:

1. change the session name, as per this PR (and add a release note to document the changing)
2. add a new attribute `services.xserver.desktopManager.session.desktopNames` and use it when generating xsession desktop files
3. override `XDG_CURRENT_DESKTOP` in `services.xserver.desktopManager.session.start` with an `export` command

Maybe the best option is 2.

symphorien commented:
There is already a breaking change for `defaultSession` in 20.03, so I would go for 1.

romildo commented:
Option 2 would be something like:

``` patch
nixos.display-managers: use new attribute for desktop names

1 file changed, 4 insertions(+), 1 deletion(-)
nixos/modules/services/x11/display-managers/default.nix | 5 ++++-

modified   nixos/modules/services/x11/display-managers/default.nix
@@ -412,6 +412,9 @@ in
             (dm: wm: let
               sessionName = "${dm.name}${optionalString (wm.name != "none") ("+" + wm.name)}";
               script = xsession dm wm;
+              desktopNames = if dm ? desktopNames
+                             then concatStringsSep ";" dm.desktopNames
+                             else sessionName;
             in
               optional (dm.name != "none" || wm.name != "none")
                 (pkgs.writeTextFile {
@@ -427,7 +430,7 @@ in
                     TryExec=${script}
                     Exec=${script}
                     Name=${sessionName}
-                    DesktopNames=${sessionName}
+                    DesktopNames=${desktopNames}
                   '';
                 } // {
                   providedSessions = [ sessionName ];
```

``` patch
nixos.xfce: set desktopNames

1 file changed, 1 insertion(+)
nixos/modules/services/x11/desktop-managers/xfce.nix | 1 +

modified   nixos/modules/services/x11/desktop-managers/xfce.nix
@@ -129,6 +129,7 @@ in
 
     services.xserver.desktopManager.session = [{
       name = "xfce";
+      desktopNames = [ "XFCE" ];
       bgSupport = true;
       start = ''
         ${pkgs.runtimeShell} ${pkgs.xfce.xfce4-session.xinitrc} &
```

romildo commented:
With option 2 it is not necessary to force the name of the session displayed by the display manager to be the same needed for `XDG_CURRENT_DESKTOP` (which is standardized).

Option 2 would allow also to use more than one desktop name per session, which is allowed by the [standard](https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html):

> If $XDG_CURRENT_DESKTOP is set then it contains a colon-separated list of strings. In order, each string is considered. If a matching entry is found in OnlyShowIn then the desktop file is shown. If an entry is found in NotShowIn then the desktop file is not shown. If none of the strings match then the default action is taken (as above). 

> $XDG_CURRENT_DESKTOP should have been set by the login manager, according to the value of the DesktopNames found in the session file. The entry in the session file has multiple values separated in the usual way: with a semicolon.

For instance, in Archlinux and Fedora there is `/usr/share/xsessions/gnome-classic.desktop` with the line `DesktopNames=GNOME-Classic;GNOME;`

worldofpeace commented:
I wish I had more time to look into this issue, I believe it is particularly important.

cc @hedning @jtojnar ?

jtojnar commented:
Yeah, option number two is what we want. It will also fix the Mate issue.

romildo commented:
I am closing this PR in favour of https://github.com/NixOS/nixpkgs/pull/83637 which implements option 2.
